core/src/main/java/com/linecorp/armeria/client/endpoint/healthcheck/HttpHealthChecker.java

```diff
@@ -89,35 +107,189 @@ private synchronized void check() {
+    }
+
+    private class HealthCheckResponseSubscriber implements Subscriber<HttpObject> {
+
+        private final ClientRequestContext reqCtx;
+        private final HttpResponse res;
+        @SuppressWarnings("NotNullFieldNotInitialized")
+        private Subscription subscription;
+        private boolean isHealthy;
+        private boolean receivedExpectedResponse;
+        private boolean updatedHealth;
+
+        @Nullable
+        private ScheduledFuture<?> pingCheckFuture;
+        private long lastPingTimeNanos;
+
+        HealthCheckResponseSubscriber(ClientRequestContext reqCtx, HttpResponse res) {
+            this.reqCtx = reqCtx;
+            this.res = res;
+        }
+
+        @Override
+        public void onSubscribe(Subscription subscription) {
+            this.subscription = subscription;
+            subscription.request(1);
+            maybeSchedulePingCheck();
+        }
+
+        @Override
+        public void onNext(HttpObject obj) {
             if (closed) {
-                return null;
+                subscription.cancel();
+                return;
             }
 
-            boolean isHealthy = false;
-            if (res != null) {
-                switch (res.status().codeClass()) {
-                    case SUCCESS:
-                        maxLongPollingSeconds = getMaxLongPollingSeconds(res);
-                        isHealthy = true;
+            try {
+                if (!(obj instanceof ResponseHeaders)) {
+                    ReferenceCountUtil.release(obj);
+                    return;
+                }
+
+                final ResponseHeaders headers = (ResponseHeaders) obj;
+                updateLongPollingSettings(headers);
+
+                final HttpStatusClass statusClass = headers.status().codeClass();
+                switch (statusClass) {
+                    case INFORMATIONAL:
+                        maybeSchedulePingCheck();
                         break;
                     case SERVER_ERROR:
-                        maxLongPollingSeconds = getMaxLongPollingSeconds(res);
+                        receivedExpectedResponse = true;
+                        break;
+                    case SUCCESS:
+                        isHealthy = true;
+                        receivedExpectedResponse = true;
                         break;
                     default:
-                        if (res.status() == HttpStatus.NOT_MODIFIED) {
-                            maxLongPollingSeconds = getMaxLongPollingSeconds(res);
+                        if (headers.status() == HttpStatus.NOT_MODIFIED) {
                             isHealthy = wasHealthy;
+                            receivedExpectedResponse = true;
                         } else {
                             // Do not use long polling on an unexpected status for safety.
                             maxLongPollingSeconds = 0;
+                            logger.warn("{} Unexpected health check response: {}", reqCtx, headers);
```

Comment:
Just noticed this is probably for #1913.
Do you think we can special case 4XX with something like "Received 4XX health check response. This generally indicates a misconfiguration of the client. Did you happen to forget to configure the endpoint group option with those of the client?"